File: engine/src/warden/git_watcher.py
"""Fetch periódico -> detecta drift (commits novos no origin não puxados).

Só fetch (read-only, não mexe em working tree). Notifica uma vez na transição
pra "atrás do origin", não a cada poll — evita spam enquanto o usuário não agiu.
"""

import threading
from collections.abc import Callable
from pathlib import Path

from warden.git import git_command, git_info


class GitWatcher:
    def __init__(
        self,
        path: Path,
        remote: str,
        interval: float,
        on_behind: Callable[[int], None],
    ):
        self.path = path
        self._remote = remote
        self._interval = interval
        self._on_behind = on_behind
        self._stop_event = threading.Event()
        self._thread: threading.Thread | None = None
        self._last_behind = 0

    def start(self) -> None:
        self._thread = threading.Thread(target=self._run, daemon=True)
        self._thread.start()

    def stop(self) -> None:
        self._stop_event.set()
        if self._thread is not None:
            self._thread.join(timeout=2.0)

    def _run(self) -> None:
        while not self._stop_event.is_set():
            self._poll_once()
            self._stop_event.wait(self._interval)

    def _poll_once(self) -> None:
        fetch = git_command(self.path, "fetch", remote=self._remote)
        if not fetch.ok:
            return  # rede instável / sem credencial — próximo poll cobre

        info = git_info(self.path)
        if info is None:
            return

        behind = info.behind or 0
        if behind > 0 and self._last_behind == 0:
            self._on_behind(behind)
        self._last_behind = behind

File: agent/Warden.Domain/Git/GitWatcher.cs
namespace Warden.Domain.Git;

/// <summary>
/// Fetch periódico → detecta drift (commits novos no origin não puxados). Só fetch (read-only, não
/// mexe em working tree). Notifica uma vez na transição pra "atrás do origin", não a cada poll —
/// evita spam enquanto o usuário não agiu. Mirror de `git_watcher.py`.
/// </summary>
public sealed class GitWatcher
{
    private readonly string _path;
    private readonly string _remote;
    private readonly TimeSpan _interval;
    private readonly Action<int> _onBehind;
    private readonly ManualResetEventSlim _stopSignal = new(false);

    private Thread? _thread;
    private int _lastBehind;

    public GitWatcher(string path, string remote, double intervalSeconds, Action<int> onBehind)
    {
        _path = path;
        _remote = remote;
        _interval = TimeSpan.FromSeconds(intervalSeconds);
        _onBehind = onBehind;
    }

    public void Start()
    {
        _thread = new Thread(Run) { IsBackground = true };
        _thread.Start();
    }

    public void Stop()
    {
        _stopSignal.Set();
        _thread?.Join(TimeSpan.FromSeconds(2));
    }

    private void Run()
    {
        while (!_stopSignal.IsSet)
        {
            PollOnce();
            _stopSignal.Wait(_interval);
        }
    }

    private void PollOnce()
    {
        var fetch = GitService.Command(_path, "fetch", _remote);
        if (!fetch.Ok) return; // rede instável / sem credencial — próximo poll cobre

        var info = GitService.Info(_path);
        if (info is null) return;

        var behind = info.Behind ?? 0;
        if (behind > 0 && _lastBehind == 0) _onBehind(behind);
        _lastBehind = behind;
    }
}
